pkg/controllers/osc/osc_controller.go

```diff
@@ -277,3 +282,65 @@ func (r *Reconciler) deleteGeneratedSecrets(ctx context.Context, md *clusterv1al
 	}
 	return nil
 }
+
+// isSupportedMachineDeployment ensures that MachineDeployment is complaint with the OperatingSystemProfile
+func (r *Reconciler) isSupportedMachineDeployment(ctx context.Context, md *clusterv1alpha1.MachineDeployment) (bool, error) {
+	ospName := md.Annotations[resources.MachineDeploymentOSPAnnotation]
+	osp := &osmv1alpha1.OperatingSystemProfile{}
+	if err := r.Get(ctx, types.NamespacedName{Name: ospName, Namespace: r.namespace}, osp); err != nil {
+		return false, fmt.Errorf("Failed to get OperatingSystemProfile: %v", err)
+	}
+
+	// Get providerConfig from machineDeployment
+	providerConfig := providerconfigtypes.Config{}
+	err := json.Unmarshal(md.Spec.Template.Spec.ProviderSpec.Value.Raw, &providerConfig)
+	if err != nil {
+		return false, fmt.Errorf("Failed to decode provider configs: %v", err)
+	}
+
+	// Ensure that OSP supports the operating system
+	if osp.Spec.OSName != osmv1alpha1.OperatingSystem(providerConfig.OperatingSystem) {
+		// Just log an error instead of returning it to avoid re-queues
```

Comment:
Well, this is why we have the terminal error convention. You should some kinda custom error that is checked right before re-queuing. So when an error is terminal we should record that as an event on the machine deployment. Users who have no access to the seed cluster wouldn't know anything about why this is failing. I would suggest to:

- Create a terminal error type 
- Decide whether the request should be re-queued or not in the reconcile function directly 
- Once the error is identified as a terminal, a recorder should record that as an event on the machine deployement.